src/cipher.rs

```diff
@@ -509,6 +511,97 @@ impl rustls_root_cert_store {
     }
 }
 
+/// A builder for a `rustls_client_cert_verifier`. This builder object can be used to configure
+/// certificate revocation lists, and then turned into a `rustls_client_cert_verifier` once ready.
+pub struct rustls_client_cert_verifier_builder {
+    _private: [u8; 0],
+}
+
+// XXX: contained value is consumed even on error, so this can contain None. but the caller
+// still needs to free it
+pub(crate) struct AllowAnyAuthenticatedClientHolder(Option<AllowAnyAuthenticatedClient>);
+
+impl CastPtr for rustls_client_cert_verifier_builder {
+    type RustType = AllowAnyAuthenticatedClientHolder;
+}
+
+impl BoxCastPtr for rustls_client_cert_verifier_builder {}
+
+impl rustls_client_cert_verifier_builder {
+    /// Create a new client certificate verifier builder for the root store.
+    ///
+    /// This copies the contents of the rustls_root_cert_store. It does not take
+    /// ownership of the pointed-to memory.
+    ///
+    /// This object can then be used to load any CRLs.
+    ///
+    /// Once that is complete, convert it into a real `rustls_client_cert_verifier`
+    /// by calling `rustls_client_cert_verifier_new()`.
+    #[no_mangle]
+    pub extern "C" fn rustls_client_cert_verifier_builder_new(
+        store: *const rustls_root_cert_store,
+    ) -> *mut rustls_client_cert_verifier_builder {
+        ffi_panic_boundary! {
+            let store: &RootCertStore = try_ref_from_ptr!(store);
+            let client_cert_verifier = AllowAnyAuthenticatedClientHolder(Some(AllowAnyAuthenticatedClient::new(store.clone())));
+            BoxCastPtr::to_mut_ptr(client_cert_verifier)
+        }
+    }
+
+    /// Add one or more certificate revocation lists (CRLs) to the client certificate verifier by
+    /// reading the CRL content from the provided buffer of PEM encoded content.
+    ///
+    /// This function may return an error if the provided buffer is not valid PEM encoded content,
+    /// or if the CRL content is invalid or unsupported.
+    #[no_mangle]
+    pub extern "C" fn rustls_client_cert_verifier_builder_add_crl(
+        verifier: *mut rustls_client_cert_verifier_builder,
+        crl_pem: *const u8,
+        crl_pem_len: size_t,
+    ) -> rustls_result {
+        ffi_panic_boundary! {
+            let client_cert_verifier_holder = match BoxCastPtr::to_box(verifier) {
+                None => {
+                    return NullParameter;
+                },
+                Some(x) => Box::leak(x), // nb. does not take ownership of the box.
```

Comment:
I think is incorrect and will cause a memory leak of the `AllowAnyAuthenticatedClientHolder` even after the contents of its options are taken.

Elsewhere we use `try_mut_from_ptr!` to produce an `&mut` reference. I think that's the right semantics here; someone else (the C code) continues to hold and own the Box, but we are mutating the contents in this function.

So that would look something like:

    let client_cert_verifier_holder: &mut AllowAnyAuthenticatedClientHolder = try_mut_ptr!(verifier)

Reply:
Thanks for catching that (I guess it's too subtle for valgrind?)

I switched over to using `try_mut_from_ptr!` throughout.